src/inmanta/agent/in_process_executor.py

```diff
@@ -68,20 +68,42 @@ def __init__(
         self.thread_pool: ThreadPoolExecutor = ThreadPoolExecutor(1, thread_name_prefix="Pool_%s" % self.name)
 
         self._cache = inmanta.agent.cache.AgentCache(self)
+        # This lock ensures cache entry can not be cleaned up when
+        # the executor is actively working and vice versa
+        self.wip_lock = asyncio.Lock()
 
         self.logger: logging.Logger = parent_logger.getChild(self.name)
 
         self._stopped = False
 
         self.failed_resource_types: FailedResourcesSet = set()
+        self.periodic_cache_cleanup_job: Optional[asyncio.Task[None]] = None
 
-    def stop(self) -> None:
+    async def start(self) -> None:
+        self.periodic_cache_cleanup_job = asyncio.create_task(self.cleanup_stale_cache_entries())
+
+    async def cleanup_stale_cache_entries(self) -> None:
+        """
+        Periodically cleans up stale entries in the cache. The clean_stale_entries
+        has to be called on the thread pool because it might call finalizers.
+        """
+
+        while not self._stopped:
+            reschedule_interval: float = 1.0
```

Comment:
I'm not sure actually, your other [comment](https://github.com/inmanta/inmanta-core/pull/7846#discussion_r1677892873) about the test being slow with a long expiry interval made me realize it would be nice to add an option to control this interval easily. I used the `is_time` validator for the option (which I had to monkey patch in the test to allow for <1s interval).
Maybe it would be less convoluted to use float everywhere then ?